lib/E2E/TensorToMemref/LowerShapedResultsToMemref.cpp

```diff
@@ -134,57 +133,45 @@ class LowerBroadcastToToLoopsPattern
 } // namespace
 
 namespace {
-class LowerLinalgGenericTensorToMemRef
-    : public OpConversionPattern<linalg::GenericOp> {
+class LowerTcpAddOp : public OpConversionPattern<tcp::AddOp> {
 public:
   using OpConversionPattern::OpConversionPattern;
   LogicalResult
-  matchAndRewrite(linalg::GenericOp op, ArrayRef<Value> operands,
+  matchAndRewrite(tcp::AddOp op, ArrayRef<Value> operands,
                   ConversionPatternRewriter &rewriter) const override {
-
-    // TODO: Replace this with more generic code operating on named
-    // structured ops too.
-
-    // These checks mirror those in BypassShapes.
-    if (!llvm::all_of(op.getOperandTypes(),
-                      [](Type type) { return type.isa<RankedTensorType>(); })) {
-      return rewriter.notifyMatchFailure(op, "all operands must be tensors");
-    }
-    if (!llvm::all_of(op.getResultTypes(),
-                      [](Type type) { return type.isa<RankedTensorType>(); })) {
-      return rewriter.notifyMatchFailure(op, "all results must be tensors");
-    }
-    if (!llvm::all_of(op.indexing_maps(), [](Attribute map) {
-          return map.cast<AffineMapAttr>().getValue().isIdentity();
-        })) {
-      return rewriter.notifyMatchFailure(
-          op, "all indexing maps must be identity maps");
-    }
-    if (!llvm::all_of(op.iterator_types(), [](Attribute str) {
-          return str.cast<StringAttr>().getValue() ==
-                 getParallelIteratorTypeName();
-        })) {
-      return rewriter.notifyMatchFailure(
-          op, "all iterator types must be 'parallel'");
-    }
-
-    SmallVector<Value, 6> memrefs(operands.begin(), operands.end());
-
     auto resultsOrFailure = allocateResults(op, rewriter, op.getLoc());
     if (failed(resultsOrFailure))
       return failure();
     auto results = *resultsOrFailure;
-    memrefs.append(results.begin(), results.end());
 
-    auto newGeneric = rewriter.create<linalg::GenericOp>(
-        op.getLoc(), llvm::None, ValueRange(memrefs), op.getAttrs());
-    newGeneric.region().getBlocks().clear();
-    BlockAndValueMapping mapper;
-    op.region().cloneInto(&newGeneric.region(), mapper);
-    for (auto memref : results) {
-      newGeneric.region().front().addArgument(
-          memref.getType().cast<MemRefType>().getElementType());
-    }
+    SmallVector<Value, 6> args;
+    args.append(operands.begin(), operands.end());
+    args.append(results.begin(), results.end());
+
+    size_t rank = op.getType().cast<RankedTensorType>().getRank();
+    SmallVector<StringRef, 6> iterators(rank, getParallelIteratorTypeName());
+    // TODO: Generalize this to other elementwise ops.
+    // All we need to do is to have a mapping of tcp.foo to scalar.foo.
+    // TODO: Should we just use linalg named ops for most of TCP?
```

Comment:
I've had the same question, fwiw.

Reply:
Yeah, I'll look into it more. If they are all this trivial, then I think we should. I'm worried about e.g. needing to add a simple helper method into the class, and then not being able to because they are generated, or having to have some sort of nested .td inside .tc madness like this:

```
ods_def<MatmulOp>:
ods_extra_declaration```
let extraClassDeclaration = [{
  ... methods ...
}];
```
def matmul(A: f32(M, K), B: f32(K, N)) -> (C: f32(M, N)) {
  C(m, n) = std_addf<k>(std_mulf(A(m, k), B(k, n)));
}

```

Or just extra verification even.